### PR TITLE
API-1448: Refacto and fix integration tests

### PR DESCRIPTION
**Pull request by willy-ahva**
**Description (for Contributor and Core Developer)**

Moving from in-memory pub-sub to pub-sub emulator.
Use the Trait in integration tests.
Reset the messenger bus after in the ProductLoader.

All the integration test files and class names were not the same. I renamed all the classes.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Connectivity/Connection/back/tests/CatalogBuilder/Enrichment/ProductLoader.php [hunk 1]
```diff
@@ -18,33 +19,27 @@
...
-    private $validator;
-
-    /** @var Client */
-    private $client;
+    private ProductBuilderInterface $builder;
+    private ObjectUpdaterInterface $updater;
+    private SaverInterface $saver;
+    private ValidatorInterface $validator;
+    private Client $client;
+    private TraceableMessageBus $messageBus;
```
Comment (tseho) [inline src/Akeneo/Connectivity/Connection/back/tests/CatalogBuilder/Enrichment/ProductLoader.php:27]: We are in 5.0, even if it's only tests, it's not always safe to change the DI typing, you never know ...

File: src/Akeneo/Connectivity/Connection/back/tests/Integration/Webhook/ProduceEventOnBulkProductCreationIntegration.php [hunk 2]
```diff
@@ -1,20 +1,23 @@
...
 use Akeneo\Test\Integration\TestCase;
+use Akeneo\Test\IntegrationTestsBundle\Messenger\AssertEventCountTrait;
 use Akeneo\Tool\Component\StorageUtils\Saver\BulkSaverInterface;
 
 /**
  * @copyright 2021 Akeneo SAS (http://www.akeneo.com)
  * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
  */
-class ProduceAPIEventOnBulkProductCreationIntegration extends TestCase
+class ProduceEventOnBulkProductCreationIntegration extends TestCase
```
Comment (tseho) [inline src/Akeneo/Connectivity/Connection/back/tests/Integration/Webhook/ProduceEventOnBulkProductCreationIntegration.php:17]: Damn, I missed that.

File: src/Akeneo/Connectivity/Connection/back/tests/Integration/Webhook/ProduceEventOnBulkProductCreationIntegration.php [hunk 3]
```diff
@@ -35,16 +38,7 @@ public function test_the_bulk_product_creation_event()
...
-
-        $envelopes = $transport->get();
-        $this->assertCount(1, $envelopes);
-
-        /** @var BulkEvent */
-        $bulkEvent = $envelopes[0]->getMessage();
-        $this->assertInstanceOf(BulkEvent::class, $bulkEvent);
-        $this->assertCount($count, $bulkEvent->getEvents());
-        $this->assertContainsOnlyInstancesOf(ProductCreated::class, $bulkEvent->getEvents());
+        $this->assertEventCount($count, ProductCreated::class);
```
Comment (tseho) [inline src/Akeneo/Connectivity/Connection/back/tests/Integration/Webhook/ProduceEventOnBulkProductCreationIntegration.php:41]: That's sooooo much cleaner :heart_eyes:

File: src/Akeneo/Connectivity/Connection/back/tests/Integration/Webhook/ProduceEventOnSingleProductUpdateIntegration.php [hunk 4]
```diff
@@ -38,26 +51,12 @@ public function test_the_single_product_update_event()
...
-
-        /** @var BulkEvent */
-        $bulkEvent = $envelopes[0]->getMessage();
-        $this->assertInstanceOf(BulkEvent::class, $bulkEvent);
-        $this->assertCount(1, $bulkEvent->getEvents());
-        $this->assertContainsOnlyInstancesOf(ProductCreated::class, $bulkEvent->getEvents());
+        $this->updater->update($product, ['enabled' => false]);
+        $constraints = $this->validator->validate($product);
+        Assert::assertCount(0, $constraints, 'The validation from the product creation failed.');
+        $this->saver->save($product);
```
Comment (tseho) [inline src/Akeneo/Connectivity/Connection/back/tests/Integration/Webhook/ProduceEventOnSingleProductUpdateIntegration.php:57]: why ?

Comment (willy-ahva) [reply src/Akeneo/Connectivity/Connection/back/tests/Integration/Webhook/ProduceEventOnSingleProductUpdateIntegration.php:57]: The goal of the scenario is to test that the service that updates and saves a product dispatches an event.
Here we were using the product loader to do it. The purpose of the product loader is to create fixtures. So firstly, we may want to use a service other than the product updater and product saver to execute the update. Secondly, as it's a fixtures purpose service, it means that it should reset the messenger bus after having been used (that's what I did).
You can see that all the other tests are using good services (savers and updaters) and not the fixtures one.

**Review by tomglvng — APPROVED** [review]

**Review by pierrejolly — APPROVED** [review]

**Review by pchasle — APPROVED** [review]

**Pull request merged**